### PR TITLE
[server] Incrementally clean historical KV state

### PR DESCRIPTION
**Pull request by luoyuxia**
### Purpose

Linked issue: #4159.

Depends on #4120. This is a stacked PR; until #4120 merges, review commit `f6914503a` only.

Historical writes retain local KV values and tombstones until their WAL offsets are covered by the lake snapshot. This change makes entries below the confirmed lake progress eligible for incremental reclamation, without waiting for the lake and local log end offsets to become equal or rebuilding the entire RocksDB instance.

Cleanup is best effort: the cleanup offset controls eligibility, while physical deletion occurs during normal RocksDB compactions. The implementation does not schedule an additional full-range compaction when the offset advances.

### Brief change log

- Preserve the previous values resolved by the initial historical lookup and reuse them during apply, so concurrent compaction cannot change merge, delete, or changelog semantics.
- Tag historical values and tombstones with the WAL offset that produced them on leader writes, pre-write-buffer flushes, and WAL recovery.
- Maintain a monotonic, exclusive historical cleanup offset from confirmed lake progress, publishing the required lake snapshot before exposing a newer offset and restoring it safely during leader recovery.
- Generalize the row-TTL compaction-filter factory so historical cleanup can use the stored WAL-offset tag and remove entries whose tag is lower than the cleanup offset.
- Decode tagged local lookup results while keeping lake values in their existing plain layout, and add tests for resolved previous values, offset tagging, cleanup boundaries, repeated progress notifications, recovery, and lookup behavior.

### Tests

- Targeted `fluss-server` unit tests: `KvValueLayoutTest`, `RowTtlCompactionFilterTest`, `HistoricalKvCompactionFilterTest`, `KvTabletTest`, and `HistoricalPartitionManagerTest` (56 tests passed).
- `./mvnw -pl fluss-server spotless:check`
- Checkstyle and RAT validation passed for the affected modules.
- `git diff --check`
- `./mvnw verify` was attempted. It stopped in `fluss-server` after 1,303 tests because the local disk was 95% full: five `ReplicaFetcherThreadTest` cases hit the 85% RocksDB disk guard, and two unrelated `ReplicaTest` local-log metric assertions failed. The historical-KV and compaction-filter tests passed in that run.

### API and Format

No public API changes.

The internal RocksDB value format for historical KV state now carries an 8-byte WAL-offset tag; a historical tombstone consists only of that tag. Lake values remain in the plain value layout.

### Documentation

No user-facing documentation changes. The design and trade-offs are described in #4159 and this PR.

### Generative AI disclosure

- [x] Yes — Codex (GPT-5), reviewed by the author.

### CODE REVIEW

**Comment by luoyuxia** [comment]
@wuchong @platinumhamburg Could you please help review this

File: fluss-server/src/main/java/org/apache/fluss/server/kv/historical/HistoricalKvTombstone.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+package org.apache.fluss.server.kv.historical;
+
+import org.apache.fluss.annotation.Internal;
+import org.apache.fluss.row.encode.KvValueLayout;
+
+import static org.apache.fluss.utils.Preconditions.checkArgument;
+
+/** Utilities for tombstones stored in local historical KV state. */
+@Internal
+public final class HistoricalKvTombstone {
```
Comment (platinumhamburg) [inline fluss-server/src/main/java/org/apache/fluss/server/kv/historical/HistoricalKvTombstone.java]: Should rename to HistoricalKvTombstones since this is a utility class and just provides static helper functions.

File: fluss-server/src/main/java/org/apache/fluss/server/kv/historical/LocalValueLookupResult.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        return Collections.unmodifiableList(primaryKeys);
+    }
+
+    /**
+     * Combines lake results with the local lookup results and creates an in-memory lookup for
+     * apply.
+     *
+     * <p>Lake values must have the same order as {@link #keysMissingLocally()}.
+     */
+    public HistoricalValueLookup createValueLookup(List<byte[]> lakeValues) {
```
Comment (platinumhamburg) [inline fluss-server/src/main/java/org/apache/fluss/server/kv/historical/LocalValueLookupResult.java:96]: keysMissingLocally() and createValueLookup() seem to have an implicit contract between them — the caller has to know to call them in the right order. Pushing that onto the caller in Replica doesn't feel like good practice.

**Review by platinumhamburg — APPROVED** [review]
LGTM +1. Only one minor suggestion.

**Comment by luoyuxia** [comment]
Let me merge it first since the risk is realtely low. If there’s any further feedback, we can address it in a follow-up PR.
